File: src/lib/overlay.ts
import { invoke, isTauri } from "@tauri-apps/api/core";

export type OverlayPosition = { x: number; y: number };

export type OverlayPayload = {
  title: string;
  body: string;
  kind?: string;
  type?: "helltide" | "legion" | "world_boss";
  bg_rgb?: number;
  scale?: number;
  bg_a?: number;
};

export type OverlayStatus = {
  supported: boolean;
  running: boolean;
  visible: boolean;
  config_mode: boolean;
  last_error: string | null;
  position: OverlayPosition | null;
};

export async function overlayStatus(): Promise<OverlayStatus | null> {
  if (!isTauri()) return null;
  try {
    return (await invoke("overlay_status")) as OverlayStatus;
  } catch {
    return null;
  }
}

export async function overlayShow(payload: OverlayPayload, position?: OverlayPosition | null): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke("overlay_show", { payload, position: position ?? null });
  } catch (e) {
    console.warn("overlay_show failed", e);
  }
}

export async function overlayHide(): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke("overlay_hide");
  } catch (e) {
    console.warn("overlay_hide failed", e);
  }
}

export async function overlayEnterConfig(position?: OverlayPosition | null): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke("overlay_enter_config", { position: position ?? null });
  } catch (e) {
    console.warn("overlay_enter_config failed", e);
  }
}

export async function overlayExitConfig(): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke("overlay_exit_config");
  } catch (e) {
    console.warn("overlay_exit_config failed", e);
  }
}

export async function overlayGetPosition(): Promise<OverlayPosition | null> {
  if (!isTauri()) return null;
  try {
    return (await invoke("overlay_get_position")) as OverlayPosition | null;
  } catch (e) {
    console.warn("overlay_get_position failed", e);
    return null;
  }
}

export async function overlaySetPosition(position: OverlayPosition): Promise<void> {
  if (!isTauri()) return;
  try {
    await invoke("overlay_set_position", { position });
  } catch (e) {
    console.warn("overlay_set_position failed", e);
  }
}

File: src-tauri/src/overlay.rs
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OverlayPosition {
  pub x: i32,
  pub y: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OverlayPayload {
  pub title: String,
  pub body: String,
  pub kind: Option<String>,
  #[serde(rename = "type")]
  pub event_type: Option<String>,
  pub bg_rgb: Option<u32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct OverlayStatus {
  pub supported: bool,
  pub running: bool,
  pub visible: bool,
  pub config_mode: bool,
  pub last_error: Option<String>,
  pub position: Option<OverlayPosition>,
}

#[cfg(windows)]
mod win32;

#[cfg(windows)]
pub use win32::OverlayManager;

#[cfg(not(windows))]
pub struct OverlayManager {
  _noop: (),
}

#[cfg(not(windows))]
impl OverlayManager {
  pub fn new() -> Self {
    Self { _noop: () }
  }

  pub fn status(&self) -> OverlayStatus {
    OverlayStatus {
      supported: false,
      running: false,
      visible: false,
      config_mode: false,
      last_error: None,
      position: None,
    }
  }

  pub fn show(&self, _payload: OverlayPayload, _position: Option<OverlayPosition>) -> Result<(), String> {
    Ok(())
  }

  pub fn hide(&self) -> Result<(), String> {
    Ok(())
  }

  pub fn enter_config(&self, _position: Option<OverlayPosition>) -> Result<(), String> {
    Ok(())
  }

  pub fn exit_config(&self) -> Result<(), String> {
    Ok(())
  }

  pub fn get_position(&self) -> Option<OverlayPosition> {
    None
  }

  pub fn set_position(&self, _pos: OverlayPosition) -> Result<(), String> {
    Ok(())
  }
}

#[derive(Clone, Default)]
pub(crate) struct Shared {
  pub toast: Arc<Mutex<Option<OverlayPayload>>>,
  pub visible: Arc<Mutex<bool>>,
  pub config_mode: Arc<Mutex<bool>>,
  pub position: Arc<Mutex<Option<OverlayPosition>>>,
  pub last_error: Arc<Mutex<Option<String>>>,
}
